app/controllers/agents_controller.rb

```diff
@@ -220,9 +220,9 @@ def redirect_back(message, options = {})
   end
 
   def build_agent
-    @agent = Agent.build_for_type(params[:agent].delete(:type),
+    @agent = Agent.build_for_type(agent_params.delete(:type),
```

Comment:
Oh yes, now I get your point! I got confused by the code as well, I assumed `.delete` would keep the original hash intact which is not the case ofc.
I don't think your suggestion would work, the first argument needs to be the value of `:type`, we could go with
``` ruby
@agent = Agent.build_for_type(agent_params[:type],
                                                    current_user,
                                                    agent_params.except(:type))
```